engine/src/main/scala/cromwell/engine/workflow/tokens/JobExecutionTokenDispenserActor.scala

```diff
@@ -0,0 +1,130 @@
+
+  /**
+    * Lazily created token pool. We only create a pool for a token type when we need it
+    */
+  var tokenPools: Map[JobExecutionTokenType, TokenPool] = Map.empty
+  var tokenAssignments: Map[ActorRef, JobExecutionToken] = Map.empty
+
+  override def receive: Actor.Receive = {
+    case JobExecutionTokenRequest(tokenType) => sendTokenRequestResult(sender, tokenType)
+    case JobExecutionTokenReturn(token) => unassign(sender, token)
+    case Terminated(terminee) => onTerminate(terminee)
+  }
+
+  private def sendTokenRequestResult(sndr: ActorRef, tokenType: JobExecutionTokenType): Unit = {
+    if(tokenAssignments.contains(sndr)) {
+      sndr ! JobExecutionTokenDispensed(tokenAssignments(sndr))
+    } else {
+      context.watch(sndr)
+      val updatedTokenPool = getTokenPool(tokenType).pop() match {
+        case TokenPoolPop(newTokenPool, Some(token)) =>
+          assignAndSendToken(sndr, token)
+          newTokenPool
+        case TokenPoolPop(sizedTokenPoolAndQueue: SizedTokenPoolAndActorQueue, None) =>
+          val (poolWithActorEnqueued, positionInQueue) = sizedTokenPoolAndQueue.enqueue(sndr)
+          sndr ! JobExecutionTokenDenied(positionInQueue)
+          poolWithActorEnqueued
+      }
+
+      tokenPools += tokenType -> updatedTokenPool
+    }
+  }
+
+  private def getTokenPool(tokenType: JobExecutionTokenType): TokenPool = tokenPools.getOrElse(tokenType, createNewPool(tokenType))
+
+  private def createNewPool(tokenType: JobExecutionTokenType): TokenPool = {
+    val newPool = TokenPool(tokenType) match {
+      case s: SizedTokenPool => SizedTokenPoolAndActorQueue(s, Queue.empty)
+      case anythingElse => anythingElse
+    }
+    tokenPools += tokenType -> newPool
+    newPool
+  }
+
+  private def assignAndSendToken(actor: ActorRef, token: JobExecutionToken) = {
+    tokenAssignments += actor -> token
+    actor ! JobExecutionTokenDispensed(token)
+  }
+
+  private def unassign(actor: ActorRef, token: JobExecutionToken) = {
+    if (tokenAssignments.contains(actor) && tokenAssignments(actor) == token) {
+      tokenAssignments -= actor
+
+      val pool = getTokenPool(token.jobExecutionTokenType) match {
+        case SizedTokenPoolAndActorQueue(innerPool, queue) if queue.nonEmpty =>
+          val (nextInLine, newQueue) = queue.dequeue
+          assignAndSendToken(nextInLine, token)
+          SizedTokenPoolAndActorQueue(innerPool, newQueue)
+        case other => other
+      }
+
+      tokenPools += token.jobExecutionTokenType -> pool.push(token)
+      context.unwatch(actor)
+    } else {
+      log.error("Job execution token returned from incorrect actor: {}", token)
+    }
+  }
+
+  private def onTerminate(terminee: ActorRef) = {
+    tokenAssignments.get(terminee) match {
+      case Some(token) =>
+        log.error("Actor {} stopped without returning its Job Execution Token ({}). Reclaiming it!", terminee, tokenAssignments(terminee))
```

Comment:
Agree. But it wasn't really adding much value so I'll just say "token". Which token was only really useful during debugging